src/Microsoft.ML.Data/Data/Conversion.cs

```diff
@@ -1130,24 +1069,23 @@ public bool TryParse(ref TX src, out U8 dst)
         /// and had only digits and the letters 'a' through 'f' or 'A' through 'F' as characters</returns>
         public bool TryParse(ref TX src, out UG dst)
         {
+            var span = src.Span;
             // REVIEW: Accomodate numeric inputs?
-            if (src.Length != 34 || src[0] != '0' || (src[1] != 'x' && src[1] != 'X'))
+            if (src.Length != 34 || span[0] != '0' || (span[1] != 'x' && span[1] != 'X'))
             {
                 dst = default(UG);
                 return false;
             }
-            int ichMin;
-            int ichLim;
-            string tx = src.GetRawUnderlyingBufferInfo(out ichMin, out ichLim);
-            int offset = ichMin + 2;
+
+            int offset = 2;
             ulong hi = 0;
             ulong num = 0;
             for (int i = 0; i < 2; ++i)
             {
                 for (int d = 0; d < 16; ++d)
                 {
                     num <<= 4;
-                    char c = tx[offset++];
+                    char c = span[offset++];
```

Comment:
Your code seems to have more CPU cycles but can you please clarify how it has fewer branches?

---
In reply to: [217328366](https://github.com/dotnet/machinelearning/pull/863#discussion_r217328366) [](ancestors = 217328366)

Reply:
> can you please clarify how it has fewer branches?

Instead of two comparisons, you have one.
`if ('0' <= c && c <= '9')` <= this has two comparisons that we `and` together.

`if ((uint)(c - '0') <= ('9'- '0'))` <= this is the suggestion, with a single comparison, and no need to `and` anything.

Reply:
I believe comparison underneath uses subtraction but will need to check.

Reply:
A subtraction is better than a branch/comparison. Also, the subtraction doesn't show up as another instruction (its part of the `lea`).

In any case, this is a nit outside the scope of this PR. I noticed it so thought it was worth a mention.

See below (24 instructions vs 20):
[sharplab snippet with disassembly](https://sharplab.io/#v2:EYLgHgbALANALiAhgZwLYB8ACAGABJgRgG4BYAKEwGZ8AmXAYVwG9zc39qAbAewDsBzXAFkA+gQAU9ABaIATrgDGASmat26ngNy8ArqlwBeXNlJl16gJYAzXOPE6LvOEvELcAWlwBybF5UAeI3EvAE4vTx8/JTVzc119dCCHJxc3CN8lU1i2AFNOZBzca1t7R2dXD28AQT9cQNsvADFw6qiY7LZ43ETbZPK01twAalwCbEz29TyCops7PtTKr0Ra+uCrFuW2sw72Lp7SlIqIleHR8azszAB2bT1LtgBfdvaqXE1BURpJGXllVR2sQ+d30RhMk3YxWCvjqRjcADJ4YpYd4wtFAbt9kkyot0n4HuZpoUoV4aiiEUi3PUmn4IbEsb0ccdBiMxhMMYT8sS5ltybhEcjqRt0bs9nputijgNeayLnT8Ld4gTnmRHkA)

Reply:
Doesn't comparison operator use subtraction?

From your suggestion:
lea ecx, [rdx-0x30]
cmp ecx, 0x9

[rdx-0x30] will incur subtraction operation and cmp operation is a subtraction operation where the result is not stored anywhere but the result is used for conditional jump from flags register. So in total you are performing 2 subtraction operations in best and worst case.

What we have right now:
cmp edx, 0x30
 jl L0016
cmp edx, 0x39
 jg L0016

Best case one subtraction operation and worst case two. May be I'm missing something but how is your proposal any better than what we have? BTW, I like this website that spits out assembly code.


---
In reply to: [218992111](https://github.com/dotnet/machinelearning/pull/863#discussion_r218992111) [](ancestors = 218992111)

Reply:
Unlike instructions like subtraction, having branches (like cmp/jumps) affects the branch predictor and hence could affect instruction pipelining (which hurts perf, if the prediction is incorrect). I am not 100% familiar with how the CPU deals with instructions, but I don't think the subtraction within the `lea` impacts cpu cycles. That instruction/pattern is really common and heavily optimized.

http://www.oopweb.com/Assembly/Documents/ArtOfAssembly/Volume/Chapter_6/CH06-1.html#HEADING1-136

Unlike the sub instruction,
> The lea instruction does not affect any of the 80x86's flag bits.

We use this 'merge range checks into one comparison' optimization within corelib.